File: sge/SGE/Math/Vec3.cpp
#include "Vec3.h"
#include "Quat.h"
#include <cassert>

namespace SGE
{
    
    const Vec3 Vec3::UNIT_X(1,0,0);
    const Vec3 Vec3::UNIT_Y(0,1,0);
    const Vec3 Vec3::UNIT_Z(0,0,1);
    const Vec3 Vec3::ZERO;
    
    Vec3::Vec3() :
        x(0),
        y(0),
        z(0)
    {
    }
    
    Vec3::Vec3(float _x, float _y, float _z) :
        x(_x),
        y(_y),
        z(_z)
    {
    }
    
    Vec3::Vec3(const Quat& q, float length)
    {
        *this = (q * Vec3::UNIT_X) * length;
    }
    
    Vec3::Vec3(const Vec3& v) :
        x(v.x),
        y(v.y),
        z(v.z)
    {
    }
    
    Vec3& Vec3::operator=(const Vec3& v)
    {
        x = v.x;
        y = v.y;
        z = v.z;
        return *this;
    }
    
    float& Vec3::operator[](int i)
    {
        assert(i >= 0 && i < 3);
        if (i == 0) 
            return x;
        else if (i == 1)
            return y;
        else
            return z;
    }
    
    float Vec3::operator[](int i) const
    {
        assert(i >= 0 && i < 3);
        if (i == 0) 
            return x;
        else if (i == 1)
            return y;
        else
            return z;
    }
    
    Vec3 Vec3::operator+(const Vec3& v) const
    {
        return Vec3(x + v.x, y + v.y, z + v.z);
    }
    
    Vec3 Vec3::operator-(const Vec3& v) const
    {
        return Vec3(x - v.x, y - v.y, z - v.z);
    }
    
    Vec3 Vec3::operator*(float f) const
    {
        return Vec3(x * f, y * f, z * f);
    }
    
    Vec3 Vec3::operator/(float f) const
    {
        assert(f != 0);
        return Vec3(x / f, y / f, z / f);
    }
    
    Vec3 Vec3::operator-() const
    {
        return Vec3(-x,-y,-z);
    }
    
    void Vec3::operator+=(const Vec3& v)
    {
        x += v.x;
        y += v.y;
        z += v.z;
    }
    
    void Vec3::operator-=(const Vec3& v)
    {
        x -= v.x;
        y -= v.y;
        z -= v.z;
    }
    
    void Vec3::operator*=(float f)
    {
        x *= f;
        y *= f;
        z *= f;
    }
    
    void Vec3::operator/=(float f)
    {
        assert(f != 0);
        x /= f;
        y /= f;
        z /= f;
    }
    
    float Vec3::length() const
    {
        return Math::sqrt(x * x + y * y + z * z);
    }
    
    Vec3 Vec3::norm() const
    {
        return *this / length();
    }
    
    float Vec3::dot(const Vec3& v) const
    {
        return x * v.x + y * v.y + z * v.z;
    }
    
    Vec3 Vec3::cross(const Vec3& v) const
    {
        return Vec3(y * v.z - z * v.y,
                    z * v.x - x * v.z,
                    x * v.y - y * v.x);
    }
    
    Vec3 Vec3::project(const Vec3& v) const
    {
        float l = length();
        assert(l != 0);
        return *this * (dot(v) / (l * l));
    }
    
    Quat Vec3::getRotTo(const Vec3& v) const
    {
        float d = dot(v);
        if (d > 0.99999)
            return Quat();
        else if (d < -0.99999)
            return Quat(makePerp(), Math::PI);
        else
            return Quat(cross(v).norm(), Math::acos(norm().dot(v.norm())));
    }
    
    Vec3 Vec3::makePerp() const
    {
        if (dot(Vec3::UNIT_Y) < 0.75)
            return cross(Vec3::UNIT_Y);
        else
            return cross(Vec3::UNIT_Z);
    }
    
    Vec3 Vec3::lerp(const Vec3& v, float i) const
    {
        return Vec3((v.x - x) * i + x,
                    (v.y - y) * i + y,
                    (v.z - z) * i + z);
    }
    
    bool Vec3::equals(const Vec3& v, float eps) const
    {
        return Math::equal(x,v.x,eps) && Math::equal(y,v.y,eps) && Math::equal(z,v.z,eps);
    }
    
    std::ostream& operator<<(std::ostream& os, const Vec3& v)
    {
        os << "<Vec3 " << v.x << " " << v.y << " " << v.z << ">";
        return os;
    }
}

File: pw3/Core/Math/Vec3.py
import math

from OpenGL.GL import *

class Vec3:
    def __init__(self, x = 0, y = 0, z = 0):
        self.x = float(x)
        self.y = float(y)
        self.z = float(z)

    def __repr__(self):
        return "Vec3(%.3f, %.3f, %.3f)" % (self.x,self.y,self.z)

    def __getitem__(self, i):
        if i == 0:
            return self.x
        elif i == 1:
            return self.y
        elif i == 2:
            return self.z
        else:
            raise IndexError

    def __setitem__(self, i, v):
        if i == 0:
            self.x = v
        elif i == 1:
            self.y = v
        elif i == 2:
            self.z = v
            
    def __eq__(self, v):
        return (self - v).length() < 1e-4
        
    def __neg__(self):
        return self * -1
        
    def __add__(self, v):
        assert isinstance(v,Vec3)
        return Vec3(self.x + v.x, self.y + v.y, self.z + v.z)

    def __sub__(self, v):
        assert isinstance(v,Vec3)
        return Vec3(self.x - v.x, self.y - v.y, self.z - v.z)

    def __mul__(self, v):
        if isinstance(v,Vec3):
            return Vec3(self.x * v.x,self.y * v.y,self.z * v.z)
        else:
            return Vec3(self.x * v, self.y * v, self.z * v)

    def __div__(self, v):
        if isinstance(v,Vec3):
            return Vec3(self.x / v.x,self.y / v.y,self.z / v.z)
        else:
            return Vec3(self.x / v, self.y / v, self.z / v)

    def __neg__(self):
        return Vec3(-self.x,-self.y,-self.z)

    def length(self):
        return math.sqrt(self.x * self.x + self.y * self.y + self.z * self.z)

    def norm(self):
        return self / self.length()

    def dot(self, v):
        return self.x * v.x + self.y * v.y + self.z * v.z

    def project(self, v):
        l = self.length()
        return self * (self.dot(v) / (l * l))

    def cross(self, v):
        return Vec3(self.y * v.z - self.z * v.y,
                    self.z * v.x - self.x * v.z,
                    self.x * v.y - self.y * v.x)

    def getRotationTo(self, v):
        d = self.dot(v)
        if d > 1 - 1e-5:
            return Quat()
        elif d < -1 + 1e-5:
            return Quat.fromAxisAngle(self.makePerp(),math.pi)
        else:
            return Quat.fromAxisAngle(self.cross(v).norm(),
                                    math.acos(self.norm().dot(v.norm())))

    def makePerp(self):
        if self.dot(Vec3.UNIT_Y) < 0.8:
            return self.cross(Vec3.UNIT_Y)
        else:
            return self.cross(Vec3.UNIT_Z)

    def lerp(self, v, l):
        return self + (v - self) * l

    def glTranslate(self):
        glTranslatef(self.x,self.y,self.z)
        
    def glScale(self):
        glScalef(self.x,self.y,self.z)
        
    def glVertex(self):
        glVertex3f(self.x,self.y,self.z)

Vec3.UNIT_X = Vec3(1,0,0)
Vec3.UNIT_Y = Vec3(0,1,0)
Vec3.UNIT_Z = Vec3(0,0,1)
Vec3.ZERO = Vec3(0,0,0)

from Quat import Quat
